GetIntoTeachingApi/Models/TeachingEvent.cs

```diff
@@ -110,5 +113,40 @@ public TeachingEvent(Entity entity, ICrmService crm, IValidatorFactory validator
             : base(entity, crm, validatorFactory)
         {
         }
+
+        protected override bool ShouldMap(ICrmService crm, OrganizationServiceContext context)
+        {
+            if (Building == null)

```

Comment:
This feels fairly complex - is there any way we can simplify it/centralise some of this logic so it could be reused? I would expect the operation to be along the lines of:

```
var existingEvent = crm.GetTeachingEvent(Id);
var removeBuilding = Building == null && existingEvent.Building != null;

if (removeBuilding) {
  // one-liner here to unlink the building, something like:
  Unlink("Building", existingEvent.Building)
}
```

The `BaseModel` could then do the heavy-lifting by inspecting the annotations on the attribute passed to infer the entity logic names etc and ultimately call `crm.DeleteLink`. I don't think you'd need the call to `context.SaveChanges()` either, as we're using the `context` passed in the call to save can be deferred (it'll be called later on once the mapping is complete anyway).